Build with latest Android Studio & keep kin-base vanilla-java

Blahartinger opened the pull request:
#### Main purpose (bug/feature/enhancement + description)
 Keeping kin-base vanilla-java instead of an Android Library. As the kin-base module is itself a fork of Stellar's java library (https://github.com/stellar/java-stellar-sdk), this feels appropriate to keep it that way.

Also updating dependencies in the project so that it will build and run with
current versions of Android tooling + Update support lib to AndroidX

#### Technical details
To accomplish this, we need to use the URI class instead of the Android
framework's Uri class. As well, we need to inject the Android version
for network request headers.
This is in the same vein, and with similar interest to what was done earlier in https://github.com/kinecosystem/kin-sdk-android/pull/23 however slimmed down to only focus on kin-base. Kik has built a Kin sdk on top of just kin-base and would like to continue to pull upstream changes from this repository. Keeping this vanilla-java in the root repo will make this process easier over time.

This updates the minimum dependencies in the project so that it will build and run with
current versions of Android tooling: Android Studio 3.4.x and gradle 5.x.
Additionally, updating to compile against the support lib to AndroidX and compileSdk = 28, which is what the support library became for API 28 onwards. targetSdk remains at 26 for now until the repo owners decide to upgrade. (however, app developers are required to target at worst the last year's sdk version so it would be prudent to update this sooner than later as well)

#### Tests (Unit/Integ)
Same as before, everything should pass

#### Documentation (Source/readme.md)

Blahartinger commented:
Travis seems to be passing aside from failed access to a private key which isn't available for public Pull Requests according to @yosriz

yosriz commented on kin-sdk/kin-sdk-lib/src/test/java/kin/sdk/TransactionSenderTest.java:
```diff
@@ -290,7 +291,7 @@ public void sendTransaction_ThirdQuery_ConnectionException() throws Exception {
     public void sendTransaction_changeTimeOut() throws Exception {
         String url = mockWebServer.url("").toString();
-        server = new Server(url, 100, TimeUnit.MILLISECONDS);
+        server = new Server(url, new OkHttpClient.Builder().build());
```
The timeout is now replaced with the default one of 10sec, this actually should have failed the test, as the test should be timeout after 500ms, but from some reason, Junit failed to timeout the test now while it's in master branch.

Blahartinger commented:
@yosriz 
"Migrating kin-base to pure java is a welcomed change, however, migrating kin-sdk to androidx is something that we don't want to do right now, as it will create backward compatibility issues as long as client developers are not migrating also to androidx."
This actually should not be the case. Android X is a completely different package structure than the support library. In fact, targeting api 26 as it does today causes issues with devs trying to use a support library that targets any other version (such as 27) since they'd have mixed versions of the support lib with the same package.

"For the same reason we don't wish to upgrade to java 8 as it may cause a compatibilities issue with older OS versions of android."
I'm fairly certain the compiler does byte weaving to support java8 features transparently but I understand your hesitation on this one and can set this back.

"KinOkHttpFactory seems redundant, why not add a new parameter for headers (header map) to the Server ctor and keep the existing timeout params? this way kin-sdk-lib can choose to add whatever it sees fits."
I'll consider changing this.

Blahartinger commented on kin-sdk/kin-base/build.gradle:
```diff
@@ -1,23 +1,17 @@
+    test.java.srcDirs += 'src/test/java'
 }
 
+sourceCompatibility = "8"
```
compiler should be byte weaving to support 7 byte I understand your hesitation and will set this back.

Blahartinger commented on kin-sdk/kin-base/src/main/java/kin/base/Server.java:
```diff
@@ -165,9 +158,7 @@ private SubmitTransactionResponse submitTransaction(Transaction transaction, Str
-        Response response = null;
-        try {
-            response = httpClient.newCall(request).execute();
+        try (Response response = httpClient.newCall(request).execute()) {
```
same as above comment, byte weaving.

Blahartinger commented:
@yosriz I've addressed all your feedback except this last one:

> KinOkHttpFactory seems redundant, why not add a new parameter for headers (header map) to the Server ctor and keep the existing timeout params? this way kin-sdk-lib can choose to add whatever it sees fits.

I don't feel that injecting the header param as a ctr param is appropriate given that it's an 'android' version. Factory generation of the OkHttp dependancy that the Server class has is a pretty normal java pattern to follow for dep constructor injection and gives callers more freedom to add additional OkHttp params.

yosriz commented on kin-sdk/kin-base/build.gradle:
```diff
@@ -1,23 +1,17 @@
+    test.java.srcDirs += 'src/test/java'
 }
 
+sourceCompatibility = "7"
```
should be 6, the try with resources is java 7 feature.

yosriz commented:
> I don't feel that injecting the header param as a ctr param is appropriate given that it's an 'android' version. Factory generation of the OkHttp dependancy that the Server class has is a pretty normal java pattern to follow for dep constructor injection and gives callers more freedom to add additional OkHttp params.

My intention was that we incline to have it as ctor params to expose those useful properties as ctor params, i.e. you can control the tx timeout, headers etc. anyway that's not critical, and having OkHttp as a param is fine.
In this case, don't you think that `KinOkHttpClientFactory` should reside in `kin-sdk` level?

>This actually should not be the case. Android X is a completely different package structure than the support library. In fact, targeting api 26 as it does today causes issues with devs trying to use a support library that targets any other version (such as 27) since they'd have mixed versions of the support lib with the same package.

We upgraded lately to the latest support (28). The problem with Android X as far as we know, that's it forces app dev to migrate the app to Android X while if app developer migrate to Android X, then combining support and AndroidX is possible using jetifier.

`kin-backup-and-restore/kin-backup-and-restore-lib/jacoco.exec`  should be removed and added to `git.ignore`

@Blahartinger see another small comment about `sourceCompatibility` of 6 instead of 7, other than these comments seem ready!

Blahartinger commented:
> In this case, don't you think that KinOkHttpClientFactory should reside in kin-sdk level?
It is in kin-sdk though, maybe I misunderood you?


> sourceCompatibility of 6 instead of 7
I'm not sure I undersatnd why java 6...andorid runs on java 7 and it doesn't look like the gradle file specifically forced it to be java 6 compat unless I missed that?

yosriz commented:
@Blahartinger 

> It is in kin-sdk though, maybe I misunderstood you?

`KinOkHttpClientFactory` resides in kin-base lib, my thinking is that it should be at `kin-sdk` lib, as it contains kin-sdk desired params.

> I'm not sure I understand why java 6...andorid runs on java 7 and it doesn't look like the gradle file specifically forced it to be java 6 compat unless I missed that?

android supports java 7 from KitKat API level 19, since our minimum api level is still 16,  we don't want to upgrade it now. (there's no real need, as kin-base doesn't use any java 7 features, that why we ask to remove the try with resources)

@Blahartinger I still need to be able to run CI with your branch, and as it's external PR, secret data is not available, do you have any idea how? I see some workarounds, maybe if you can give me access to your PR to add commits , I can add my commit and it will help, another option is that I'll open a dedicated branch so you can direct your PR there and not directly to master

Blahartinger commented:
@yosriz invited you as a collaborator to the fork.
